### PR TITLE
Added forward derivatives for neg, diag, inverse, linalg_eig

### PR DESCRIPTION
**Pull request by mattarroz**
Recreated due to CI failures as per comment https://github.com/pytorch/pytorch/pull/67339#issuecomment-959893293

===

See also discussion in #10223, starting from [this](https://github.com/pytorch/pytorch/issues/10223#issuecomment-949499666) comment

The formulas for the derivatives are taken from https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf.

As indicated, the method linalg_eig_jvp should be used instead of linalg_eig_jvp_eigenvalues and linalg_eig_jvp_eigenvectors in the future. Due to a codegen limitation, this is not yet possible.

CC @albanD @Lezcano

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:2523]: While the paper does it this way for notational convenience, it's better to just divide later when you use it.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:2527]: Might be simpler to write `*` rather than `at::mul` and `/` rather than `at::div` for readability.

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2527]: A note on multiplying by `Fconj` / `eye`.

You might want to look how it's done in `linalg_eig_backward` and the tricks that are used there to avoid having to use mathematical tricks, but rather use advanced indexing. The problem with using masks, besides the efficiency, is that `NaN * 0 = NaN`, so you might be  propagating NaNs that you created in your intermediate computations, which is annoying. For example, isn't the first return equal to `dVfactor.diagonal(0, -2, -1)`?

https://github.com/pytorch/pytorch/blob/04fe4382ec8e6031d353530d6b8446ac9c9f507e/torch/csrc/autograd/FunctionsManual.cpp#L2433-L2492

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:2507]: A better way of doing this might be the following
```suggestion
  const auto dAComplex = dA.to(c10::toComplexType(dA.scalar_type()));
```
Also, doesn't the current way assume that `dA` is real? This may not be the case if the input matrix is complex. In any case, the proposed change handles this case.

Same in the other places.

**Review by lezcano — COMMENTED** [review]
Sorry for coming late to the party!

I left a few comments.

As the code is repeated three times, the changes should be applied in the 3 places. Even better, it would be to have the two auxiliary functions just call the main function that returns the pair and `std::get` the appropriate return.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mattarroz) [reply torch/csrc/autograd/FunctionsManual.cpp:2507]: Good catch :-) I used a real matrix in my example, so I would not notice if it only worked for real dA

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2507]: Weird that the tests didn't catch this one... cc @albanD

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mattarroz) [reply torch/csrc/autograd/FunctionsManual.cpp:2523]: > Sorry for coming late to the party!
> 
> I left a few comments.
> 
> As the code is repeated three times, the changes should be applied in the 3 places. Even better, it would be to have the two auxiliary functions just call the main function that returns the pair and `std::get` the appropriate return.

I was thinking about factoring the code like that too, but then I thought "this is only provisional until the codegen limitation is fixed". But now it seems like we use the separate calls for eigenvalues and eigenvectors. Anyway, I'll fix it like you suggested it.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2507]: Nevermind, the tests are red, so it could potentially be caught by the tests only that those reds are shadowing this.

On another note, @anjali411 it would be cool to have a shortcut to do this, casting a real matrix to its complex counterpart taking care of whether it's half / single / double precision.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2523]: For what is worth, we're working on always enforcing to compute the `jvp` for all the multiple outputs at once and returning them in a tuple, so having just one function and then using the `std::get`s is the best future-proof solution :)

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mattarroz) [reply torch/csrc/autograd/FunctionsManual.cpp]: Yes indeed, it is equal to `dVfactor.diagonal(0, -2, -1)`. Fixed it :-)

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mattarroz) [reply torch/csrc/autograd/FunctionsManual.cpp]: BTW, I still get the error from test_forward_mode_AD_corrcoef_cpu_complex128, test_forward_mode_AD_corrcoef_cpu_float64. I did not work on the correlation coefficients, so is this unrelated or could it be a side effect?

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp]: I think that `corrcoef` might be a composite function of this function, so it may so happen that you need to activate forward AD for these function. Just add the pertinent flag to their `OpInfo`.

**Review by lezcano — APPROVED** [review]
What a nice and compact final solution!
I'm approving this modulo making the CI happy.

**Comment by mattarroz** [comment]
> What a nice and compact final solution!

Thanks for your input! However, I think there was a mistake in that solution... there was an erroneous dVFactor (already removed, see [fixup commit](https://github.com/pytorch/pytorch/pull/67837/commits/80d8189c4b2acb7c989cedbab1813150d74ad0a7)).

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mattarroz) [reply torch/csrc/autograd/FunctionsManual.cpp]: Added the `supports_forward_ad` flag, now it is working :+1:

**Comment by albanD** [comment]
Master is a bit broken right now. It should be back to green in 10min.

**Review by albanD — COMMENTED** [review]
LGTM
Waiting for CI signal.
If you're not sure if master is good to rebase on, you can use https://hud.pytorch.org/ to check the current status (each line is a commit, each column a different build).

**Review by albanD — APPROVED** [review]

File: tools/autograd/derivatives.yaml
Comment (Lezcano) [inline tools/autograd/derivatives.yaml:1396]: nit. you could simply do this and remove the auxiliary functions
```suggestion
  eigenvalues: std::get<0>(linalg_eig_jvp(self_t, eigenvalues, eigenvectors))
  eigenvectors: std::get<1>(linalg_eig_jvp(self_t, eigenvalues, eigenvectors))
```

**Review by lezcano — COMMENTED** [review]
Left a very small comment, given that you'll need to update the PR as there's a merge conflict. Feel free to add it (or not, no big deal) when you resolve the merge conflict / CI reds.

File: tools/autograd/derivatives.yaml
Comment (mattarroz) [reply tools/autograd/derivatives.yaml]: That's better for sure :-) Fixed it.

**Comment by Lezcano** [comment]
I think that after the rebase, some pesky submodule got included in the PR. Remember to do `git submodule update --init --recursive` after rebasing your branch to avoid this.
Could you remove these updates from this PR?

After this, this PR should be ready to merge. cc @soulitzer as @albanD is on PTO afaik.

**Pull request closed without merging**
